File: benchmarks/academic/programs/test2_statistics.ts
// Test 2: Statistical Analysis — Descriptive stats on time-series data
// Real-world analog: Monitoring system computing latency percentiles,
//   analytics dashboard for e-commerce transaction values
// Data: Simulated API response latencies (log-normal-like distribution)

function statistics(size: number): string {
    let data: number[] = [];
    let i: number = 1;
    while (i <= size) {
        // Simulate API latency: base ~50ms + spikes + noise
        // Log-normal-like: exp(sin) produces right-skewed distribution
        const base: number = 50.0;
        const spike: number = Math.abs(Math.sin(i * 0.0073) * Math.cos(i * 0.0031)) * 200;
        const jitter: number = Math.sin(i * 3.17) * 10;
        data.push(base + spike + jitter);
        i = i + 1;
    }

    // Pass 1: Mean
    let sum: number = 0;
    data.forEach((v: number) => { sum = sum + v; });
    const mean: number = sum / size;

    // Pass 2: Variance + min/max
    let variance_sum: number = 0;
    let min_val: number = data[0];
    let max_val: number = data[0];
    let above_threshold: number = 0;
    data.forEach((v: number) => {
        const diff: number = v - mean;
        variance_sum = variance_sum + diff * diff;
        if (v < min_val) { min_val = v; }
        if (v > max_val) { max_val = v; }
        if (v > 150) { above_threshold = above_threshold + 1; }
    });
    const std_dev: number = Math.sqrt(variance_sum / size);

    // Pass 3: Count within 1 std dev (normality check)
    let within_1sd: number = 0;
    data.forEach((v: number) => {
        if (v > mean - std_dev) {
            if (v < mean + std_dev) {
                within_1sd = within_1sd + 1;
            }
        }
    });

    const pct_within: number = Math.floor(within_1sd * 100 / size);
    return Math.floor(mean).toString() + "," + Math.floor(std_dev).toString() + "," + pct_within.toString();
}

function run(): void {
    console.log(statistics(2000000));
}
run();

File: benchmarks/academic/rust_handwritten/test2_statistics.rs
// Hand-optimized Rust: Statistics 2M
// Key optimizations vs Tyrus:
// 1. Vec::with_capacity (pre-allocate)
// 2. for loop with usize (vs while with f64)
// 3. Single-pass stats where possible
// 4. Direct iteration (no .iter().cloned())

fn statistics(size: usize) -> String {
    let mut data = Vec::with_capacity(size);
    for i in 0..size {
        let fi = i as f64;
        data.push((fi * 0.001).sin() * 100.0 + (fi * 0.002).cos() * 50.0);
    }

    // Pass 1: sum
    let sum: f64 = data.iter().sum();
    let mean = sum / size as f64;

    // Pass 2: variance + min/max + threshold count
    let mut variance_sum = 0.0_f64;
    let mut above_threshold = 0_usize;
    for &v in &data {
        let diff = v - mean;
        variance_sum += diff * diff;
        if v > 150.0 { above_threshold += 1; }
    }
    let std_dev = (variance_sum / size as f64).sqrt();

    // Pass 3: within 1 std dev
    let within_1sd = data.iter()
        .filter(|&&v| v > mean - std_dev && v < mean + std_dev)
        .count();

    let pct_within = (within_1sd * 100) / size;
    format!("{},{},{}", mean.floor() as i64, std_dev.floor() as i64, pct_within)
}

fn main() {
    println!("{}", statistics(2_000_000));
}
